python/pyiceberg/expressions/base.py

```diff
@@ -342,107 +349,104 @@ def __str__(self) -> str:
+    inverse: ClassVar[type]
+    as_bound: ClassVar[type]
+    term: UnboundTerm[T]
 
-    def bind(self, schema: Schema, case_sensitive: bool) -> BoundIsNull[T]:
-        bound_ref = self.term.bind(schema, case_sensitive)
-        return BoundIsNull(bound_ref)
+    def __invert__(self) -> inverse:
+        return self.inverse(self.term)
 
+    def bind(self, schema: Schema, case_sensitive: bool = True) -> BooleanExpression:
+        bound_term = self.term.bind(schema, case_sensitive)
+        return self.as_bound(bound_term)
 
-class BoundIsNull(BoundPredicate[T]):
-    def __invert__(self) -> BoundNotNull:
-        return BoundNotNull(self.term)
 
-    def _validate_literals(self):  # pylint: disable=W0238
-        if self.literals:
-            raise AttributeError("Null is a unary predicate and takes no Literals.")
+@dataclass(frozen=True)
+class BoundUnaryPredicate(Bound[T], BooleanExpression):
+    inverse: ClassVar[type]
+    term: BoundTerm[T]
 
+    def __invert__(self) -> inverse:
+        return self.inverse(self.term)
 
-class NotNull(UnboundPredicate[T]):
-    def __invert__(self) -> IsNull:
-        return IsNull(self.term)
 
-    def _validate_literals(self):  # pylint: disable=W0238
-        if self.literals:
-            raise AttributeError("NotNull is a unary predicate and takes no Literals.")
+class IsNull(UnaryPredicate[T]):
+    pass
 
-    def bind(self, schema: Schema, case_sensitive: bool) -> BoundNotNull[T]:
-        bound_ref = self.term.bind(schema, case_sensitive)
-        return BoundNotNull(bound_ref)
 
+class NotNull(UnaryPredicate[T]):
+    pass
 
-class BoundNotNull(BoundPredicate[T]):
-    def __invert__(self) -> BoundIsNull:
-        return BoundIsNull(self.term)
 
-    def _validate_literals(self):  # pylint: disable=W0238
-        if self.literals:
-            raise AttributeError("NotNull is a unary predicate and takes no Literals.")
+class BoundIsNull(BoundUnaryPredicate[T]):
+    def __new__(cls, term: BoundTerm[T]):
+        if term.ref().field.required:
+            return AlwaysFalse()
+        return super().__new__(cls)
 
 
-class IsNaN(UnboundPredicate[T]):
-    def __invert__(self) -> NotNaN:
-        return NotNaN(self.term)
+class BoundNotNull(BoundUnaryPredicate[T]):
+    def __new__(cls, term: BoundTerm[T]):
+        if term.ref().field.required:
+            return AlwaysTrue()
+        return super().__new__(cls)
 
-    def _validate_literals(self):  # pylint: disable=W0238
-        if self.literals:
-            raise AttributeError("IsNaN is a unary predicate and takes no Literals.")
 
-    def bind(self, schema: Schema, case_sensitive: bool) -> BoundIsNaN[T]:
-        bound_ref = self.term.bind(schema, case_sensitive)
-        return BoundIsNaN(bound_ref)
+IsNull.inverse = NotNull
+IsNull.as_bound = BoundIsNull
+NotNull.inverse = IsNull
+NotNull.as_bound = BoundNotNull
+BoundIsNull.inverse = BoundNotNull
+BoundNotNull.inverse = BoundIsNull
```

Comment:
This was my main goal because I think it is easy to validate that the behavior is correct. Reimplementing `__inverse__` etc. in every class seemed like it was going to cause trouble later on.

Reply:
With the other refactoring, I think it was reasonable to put the `__invert__` implementations back in since they're one-line changes. Thanks, @Fokko!